src/server/mod.rs

```diff
@@ -162,10 +163,29 @@ pub fn find_stdlib_signatures(
         })
 }
 
+#[derive(Default)]
+struct LspServerState {
+    buckets: Vec<String>,
+}
+
+impl LspServerState {
+    // XXX: rockstar (21 Jun 2022) - This `allow` pragma is temporary, until we can add
+    // bucket completion, which is blocked on the completion refactor.
+    #[allow(dead_code)]
+    pub fn buckets(&self) -> &Vec<String> {
+        &self.buckets
+    }
+
+    pub fn set_buckets(&mut self, buckets: Vec<String>) {
+        self.buckets = buckets;
+    }
+}
+
 pub struct LspServer {
     client: Arc<Mutex<Option<Client>>>,
     diagnostics: Vec<Diagnostic>,
     store: store::Store,
+    state: Mutex<RefCell<LspServerState>>,
```

Comment:
`LspServer` implements the trait `lspower::LanguageServer`. The functions that the trait exposes all take a `&self` as the first argument. When internal mutability is needed, e.g. the server state, you wrap that type in a `RefCell` so it can be mutated in a function that only has `&self` rather than `&mut self`.

The `Mutex` is because we need thread safety for that reference. I wanted to use `RwLock` (and will likely refactor this to look more like `Store` and use `RwLock`), but `RwLock<T>` is only `Send + Sync` when `T` is `Send + Sync` and...if you're still reading, you didn't bail on The Hard Thing™ before I did. It might be as easy as just using the marker traits on `LspServerState`, but planned refactor makes this kind of a non-issue.

Reply:
> Plus I was able to locally remove the RefCell, and get it to compile and tests passing.

Huh. I feel genuinely gaslit now. I would not have reached for `RefCell` if I didn't find a case where I needed it. Seems we can get away without it.